Fixes part of #123, #116, and #130: Cellular Data Controller

jamesxu0 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of #123, #116, and #130. Adds a cellular data controller to store and retrieve a preference on whether to show the CellularDataFragment.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

jamesxu0 committed:
Created CellularDialogController

jamesxu0 committed:
Started making tests for controller

jamesxu0 committed:
Removed import

jamesxu0 committed:
Fixed test cases and renamed to hideDialog

jamesxu0 committed:
Fixed comments

BenHenning commented:
This is fixing parts of #123, #116, and #130.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -5,22 +5,32 @@ import android.os.Bundle
   lateinit var stateFragmentPresenter: StateFragmentPresenter
-  // Control this boolean value from controllers in domain module.
-  private var showCellularDataDialog = true
+  private var hideCellularDataDialog = false
```
Is this name descriptive of what it's used for? It seems to me that this is more for whether we should avoid showing the dialog.

Maybe keep ``useCellularDataDialog`` and invert the retrieved value when you initialize it.

jamesxu0 replied:
We would only want to invert the value the very first time the default value false is read though. How would we know it's the very first time?

BenHenning replied:
Is this still true after my proto comment below? I would guess that whether we want to show the dialog is invariant of whether it's a default value (we should assume that we always show the cellular dialog by initializing ``useCellularDataDialog`` to true and then overwrite once the cache is read). Or am I maybe missing something?

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/CellularDialogController.kt:
```diff
@@ -0,0 +1,48 @@
+  private val cellularDataStore = cacheStoreFactory.create("cellular_data_preference", CellularDataPreference.getDefaultInstance())
+
+  init {
+    // Prime the cache ahead of time so that any existing history is read prior to any calls to setHideDialogPreference().
```
Is this actually needed? I know why we do it for the user app history controller, but it doesn't seem necessary for this controller. Maybe omit this entire init block?

jamesxu0 replied:
Oh I thought this was needed to be able to read the stored value of the preference before making any calls to setHideDialogPreference. Why isn't it needed here?

BenHenning replied:
Since setHideDialogPreference requires a user action (clicking a checkbox & closing a dialog) it's really unlikely that users can bring up the dialog & save the setting before the cache is read. The cache will still be read just by adding an observer to it.

The reason the app history one needs to be primed is because a lifecycle method overrides the state rather than a user action. This means that it's quite likely we mark the app as opened long before the UI receives the state. This would result in the app always showing "Welcome back to Oppia" even on the first run. Priming the cache ensures that the first result observed by the UI is always the true first value and not a potentially overwritten value. Given that we rely on a user action to change the store here, this race condition is effectively impossible to run into.

BenHenning commented on domain/src/test/java/org/oppia/domain/audio/CellularDialogControllerTest.kt:
```diff
@@ -0,0 +1,203 @@
+    setUpTestApplicationComponent()
+    testBlockingDispatcher.pauseDispatcher()
+  }
+
```
Nit: please remove unnecessary extra newline.

BenHenning commented on domain/src/test/java/org/oppia/domain/audio/CellularDialogControllerTest.kt:
```diff
@@ -0,0 +1,203 @@
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testController_providesInitialLiveData_indicatesNotHideDialog()
```
Maybe: ``indicatesToNotHideDialog``. Ditto elsewhere.

BenHenning commented on domain/src/test/java/org/oppia/domain/audio/CellularDialogControllerTest.kt:
```diff
@@ -0,0 +1,203 @@
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testController_setHidePrefFalse_providesLiveData_indicatesNotHideDialog()
```
Suggest also adding a test that demonstrates a new controller & observer properly retrieves the on-disk state from a previously set preference. You can simulate a new controller by calling setUpTestApplicationComponent() a second time (and can verify via step debugging that the instance of the cellular dialog controller is actually a different object during the second half of the test). This verifies the "restart app" functionality.

BenHenning commented on domain/src/test/java/org/oppia/domain/audio/CellularDialogControllerTest.kt:
```diff
@@ -0,0 +1,203 @@
+  @ObsoleteCoroutinesApi
+  fun setUp() {
+    setUpTestApplicationComponent()
+    testBlockingDispatcher.pauseDispatcher()
```
Maybe add a brief comment explaining why a separate dispatcher is needed for this.

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -127,3 +127,7 @@ message ChapterProgress {
 }
+
+message CellularDataPreference {
+  bool hideDialog = 1;
```
Please add comments for the message & field to explain what it's for.

Proto field names should be in snake_case.

Regarding the name, it made me think that this structure might be incomplete. Don't we need to store two things:
- Whether to show the dialog again?
- In the case where we shouldn't show the dialog, whether to play audio on cellular?

jamesxu0 replied:
Yeah I think it's needed too.

BenHenning reviewed (changes requested):
Thanks @jamesxu0! Structurally this looks great, but I have some concerns about whether we're modeling the right things here. See my comment on the proto change for context.

jamesxu0 committed:
Added use cellular data value and added test cases

jamesxu0 commented:
@BenHenning PTAL. Added a use_cellular_data preference that is used to determine whether to play audio if you don't want to show the dialog again. Also, updated test cases.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -52,12 +65,12 @@ class StateFragment : InjectableFragment(), CellularDataInterface {
-    // saveUserChoice -> true -> save this preference
-    // saveUserChoice -> false -> do not save this preference
+    if (saveUserChoice) cellularDialogController.setHideDialogPreference(true)
+    cellularDialogController.setUseCellularDataPreference(true)
```
Shouldn't we only be changing this value permanently if the user asks to not show the dialog again? Otherwise we should always re-ask.

Similarly, shouldn't we be setting useCellularData here, and only setting useCellularData in the observer callback above if we shouldn't show the cellular dialog again?

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/CellularDialogController.kt:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  /**
+   * Returns a [LiveData] result indicating whether the user wants to hide the CellularDataDialog. This is guaranteed to
```
Nit on documentation: I don't think this guarantee is made anymore because we are no longer priming the cache. However, I think this adjustment is okay--let's just remove this detail from the documentation.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragment.kt:
```diff
@@ -5,22 +5,35 @@ import android.os.Bundle
-    // TODO(#116): Code to control the value of showCellularDataDialog using AudioController.
+    cellularDialogController.getCellularDataPreference()
+      .observe(this, Observer<AsyncResult<CellularDataPreference>>{
+      if (it.isSuccess()) {
```
I think it would also be nice to add test cases for StateFragment if possible to ensure the dialog is shown at the right times, and that the persistent setting works as expected.

BenHenning reviewed (changes requested):
Thanks @jamesxu0! I'm hoping there are a few UI tests that can be added and I have one concern with saving preferences, but otherwise the PR looks good to me.

jamesxu0 committed:
Started adding test cases

jamesxu0 committed:
Setup initial testing environment for state fragment

jamesxu0 committed:
Add more test cases

jamesxu0 commented:
@BenHenning PTAL. Added a suite of test cases for StateFragment.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivityPresenter.kt:
```diff
@@ -0,0 +1,21 @@
+    activity.setContentView(R.layout.state_fragment_test_activity)
+    activity.supportFragmentManager.beginTransaction().add(
+      R.id.state_fragment_placeholder,
+      StateFragment()
```
You may want to check whether the fragment is already added to avoid readding it upon config changes.

BenHenning commented on app/src/main/res/layout/state_fragment_test_activity.xml:
```diff
@@ -0,0 +1,8 @@
+  android:id="@+id/state_fragment_placeholder"
+  android:layout_width="match_parent"
+  android:layout_height="match_parent"
+  tools:context=".home.HomeActivity" />
```
Is this context correct?

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt:
```diff
@@ -0,0 +1,193 @@
+import org.oppia.util.threading.BlockingDispatcher
+import javax.inject.Singleton
+
+/** Tests for [ExplorationActivity]. */
```
Please update the component name being tested here--pretty sure it isn't ``ExplorationActivity``

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/CellularDialogController.kt:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  /** Saves that the user's preference on whether to use cellular data. */
+  fun setUseCellularDataPreference(useData: Boolean) {
```
Could this be combined with the above? I suspect we only ever want to save the user's preference if they don't want to show the dialog again. Perhaps:

``setNeverUseCellularDataPreference()``
``setAlwaysUseCellularDataPreference()``

(these would call a common private helper that always sets the hide dialog bit to true and sets the corresponding cellular preference). Longer term we could have a third method to reset the data source if we add a preference for this in the user settings menu.

BenHenning reviewed (approved):
Thanks @jamesxu0 --excellent tests! This generally looks good. Can you also record a few videos in Android Studio with your emulator showing the different flows? That would be nice to help provide visual context for this work.

Feel free to submit once comments are resolved, or re-assign to me if you want me to do one more pass. Generally I think this looks good though.

jamesxu0 committed:
Changed cellular data controller methods to always and never use cellular data

jamesxu0 committed:
Merge branch 'develop' into cellular-data-controller

rt4914 commented:
@BenHenning @jamesxu0 
This PR has one `testing` folder here: https://github.com/oppia/oppia-android/tree/develop/app/src/main/java/org/oppia/app/player/state

Do we actually need this?

jamesxu0 commented:
> @BenHenning @jamesxu0
> This PR has one `testing` folder here: https://github.com/oppia/oppia-android/tree/develop/app/src/main/java/org/oppia/app/player/state
> 
> Do we actually need this?

@rt4914 @BenHenning I made a testing folder to contain the activity and presenter that is used in the espresso StateFragment tests. Ben said he ran into a lot of problems using FragmentScenario, so I created a separate activity to start StateFragment for testing purposes.
